### PR TITLE
Solve issue #8014

### PR DESCRIPTION
**Pull request by LaurentBerger**
Solve issue #8014 
I made a program to test this PR https://gist.github.com/LaurentBerger/541592d4b921483dae722e24560a8667

mean time (ms) all png images in folder samples/data
Foreground             Background              back&Foreground       #pixels
3.98                        3.78                        10.44                         190653    without stat
7.22                        7.12                        15.72                         190653    with stat

cc

fig2ccstat

### CODE REVIEW

File: modules/imgproc/src/connectedcomponents.cpp [hunk 1]
```diff
@@ -1737,14 +1755,40 @@ int cv::connectedComponentsWithStats(InputArray _img, OutputArray _labels, Outpu
     _labels.create(img.size(), CV_MAT_DEPTH(ltype));
     cv::Mat labels = _labels.getMat();
     connectedcomponents::CCStatsOp sop(statsv, centroids);
-    if (ltype == CV_16U){
-        return connectedComponents_sub1(img, labels, connectivity, ccltype, sop);
+    int nbComponents;
+    if (ltype == CV_16U) {
```
Comment (sovrasov) [inline modules/imgproc/src/connectedcomponents.cpp]: The first two branches can be joined. Also it's seems that 8U labels are [supported](https://github.com/opencv/opencv/blob/dcbed8d676a4f0879fc25c31aeaf22bd738f2c63/modules/imgproc/src/connectedcomponents.cpp#L1670). Could you check this?

File: modules/imgproc/src/connectedcomponents.cpp [hunk 2]
```diff
@@ -1737,14 +1755,40 @@ int cv::connectedComponentsWithStats(InputArray _img, OutputArray _labels, Outpu
...
         return 0;
     }
+    if (!(ccltype & CCL_FORE_BACK_GROUND))
+        return nbComponents;
+    bitwise_not(img, img);
+    Mat labelsnot(img.size(), CV_MAT_DEPTH(ltype));
+    Mat statsv2, centroids2;
+    connectedcomponents::CCStatsOp sop2(statsv2, centroids2);
+    int nbComponentsNot=0;
+    if (ltype == CV_16U) {
```
Comment (sovrasov) [inline modules/imgproc/src/connectedcomponents.cpp]: The same thing with this `if`.

File: modules/imgproc/src/connectedcomponents.cpp [hunk 3]
```diff
@@ -1737,14 +1755,40 @@ int cv::connectedComponentsWithStats(InputArray _img, OutputArray _labels, Outpu
...
+        nbComponents = connectedComponents_sub1(img, labels, connectivity, ccltype&(~CCL_FORE_BACK_GROUND), sop);
     }
-    else{
+    else {
         CV_Error(CV_StsUnsupportedFormat, "the type of labels must be 16u or 32s");
         return 0;
     }
+    if (!(ccltype & CCL_FORE_BACK_GROUND))
+        return nbComponents;
+    bitwise_not(img, img);
```
Comment (sovrasov) [inline modules/imgproc/src/connectedcomponents.cpp]: Now function modifies the input image in `CCL_FORE_BACK_GROUND` mode, so sufficient note should be added to documentation.

Comment (LaurentBerger) [inline modules/imgproc/src/connectedcomponents.cpp]: connectedComponentsWithStat using   CV_8U fails  : 

> OpenCV Error: Assertion failed ((unsigned)i0 < (unsigned)size.p[0]) in cv::Mat::at, file G:\Lib\opencv\modules\core\include\opencv2/core/mat.inl.hpp, line 955

I must dig in the code...

Comment (alalek) [inline modules/imgproc/src/connectedcomponents.cpp]: BTW, There was a [fix](https://github.com/opencv/opencv/pull/8379/files) for this line with assertion.

Comment (LaurentBerger) [inline modules/imgproc/src/connectedcomponents.cpp]: I updated my repo so now this fix is in my code.  But exception is still here with this code : 
```
         Mat img = imread("f:/lib/opencv/samples/data/box.png", IMREAD_GRAYSCALE),imgThresh, imgThreshNot;
        Mat labelsBackground;
        Mat statsvBackground, centroidBackground;

        double thresh = threshold(img, imgThresh, 50, 255, THRESH_OTSU);
        bitwise_not(imgThresh, imgThreshNot);
        int nbLabelsBackGround = connectedComponentsWithStats(imgThreshNot, labelsBackground, statsvBackground, centroidBackground, 4, CV_8U, CCL_WU);

```
I thnik problem is at  [this line ](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/connectedcomponents.cpp#L207) LabelT is uchar and [at line](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/connectedcomponents.cpp#L207) there is no check lunique is equal to 255. Then[ sop dimension](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/connectedcomponents.cpp#L207) is wrong  and exception. Bug should exist for CV_16U and CV_32U . May be some CV_ASSERT should be inserted

**Pull request closed without merging**